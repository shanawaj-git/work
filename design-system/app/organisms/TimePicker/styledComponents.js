import styled from 'styled-components';
import { colorSelector } from 'themes';

export const TimeContainer = styled.div`
  display: flex;
  width: 101px;
  height: 52.5px;
  align-items: center;
  justify-content: center;
  border-radius: 12px;
  ${({ selected }) =>
    selected &&
    `
    background-color: #12a1c0;
  `}
  ${({ disabled }) =>
    disabled &&
    `
     pointer-events: none;
  `}

  > p.MuiTypography-root {
    color: ${colorSelector('p400')};
    font-style: normal;
    font-weight: 500;
    font-size: 15px;
    line-height: 22px;
    letter-spacing: 0.3px;
    text-align: center;
    ${({ selected }) =>
      selected &&
      `
    font-weight: 600;
    color: ${colorSelector('c200')};
  `}
    ${({ disabled }) =>
      disabled &&
      `
    color: ${colorSelector('c600')};
  `}
  }
`;
