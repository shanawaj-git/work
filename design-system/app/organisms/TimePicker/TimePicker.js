import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';

import CollapsibleFieldContainer from 'molecules/CollapsibleFieldContainer';

import timeIcon from 'images/time-icon.svg';
import Typography from 'atoms/Typography';
import { TimeContainer } from './styledComponents';

const getDate = timeString => new Date(`1970-01-01T${timeString}:00Z`);
const padStartWithZero = num => String(num).padStart(2, '0');
const getTimeString = date =>
  `${padStartWithZero(date.getUTCHours())}:${padStartWithZero(
    date.getUTCMinutes(),
  )}`;
const getLabelString = timeString => timeString.replace(/^0/, '');
const getCellData = ({ time, value, shouldDisableTime }) => {
  const cellValue = getTimeString(time);
  return {
    value: cellValue,
    label: getLabelString(cellValue),
    isSelected: cellValue === value,
    isDisabled: shouldDisableTime && shouldDisableTime(cellValue) === true,
  };
};
const generateCellsData = ({
  value,
  fromTime,
  toTime,
  interval,
  shouldDisableTime,
}) => {
  const startTime = getDate(fromTime);
  const endTime = getDate(toTime);
  const time = startTime;
  const cellsData = [];
  do {
    cellsData.push(
      getCellData({
        time,
        value,
        shouldDisableTime,
      }),
    );
    time.setTime(time.getTime() + getDate(interval).getTime());
  } while (time.getTime() <= endTime.getTime());

  return cellsData;
};

const getCellContainerClassNames = (selected, disabled) => {
  const classNames = ['timepicker-time-cell'];
  if (selected) {
    classNames.push('selected');
  }
  if (disabled) {
    classNames.push('disabled');
  }
  return classNames.join(' ');
};

const Picker = props => {
  const {
    value,
    fromTime,
    toTime,
    interval,
    onChange,
    shouldDisableTime,
  } = props;

  const [cellsData, setCellsData] = useState([]);
  useEffect(() => {
    setCellsData(
      generateCellsData({
        value,
        fromTime,
        toTime,
        interval,
        shouldDisableTime,
      }),
    );
  }, [value, fromTime, toTime, interval, shouldDisableTime]);

  const onCellClick = cellValue => onChange && onChange(cellValue);

  return (
    <div className="timepicker-container p3 grid grid-cols-3 md:grid-cols-4 lg:grid-cols-6 xl:grid-cols-8 2xl:grid-cols-10">
      {cellsData.map(({ value: cellValue, isSelected, label, isDisabled }) => (
        <TimeContainer
          className={getCellContainerClassNames(isSelected, isDisabled)}
          key={cellValue}
          selected={isSelected}
          disabled={isDisabled}
          onClick={() => !isDisabled && onCellClick(cellValue)}
          data-testid={`div-timepicker-cell-container-${cellValue}`}
        >
          <Typography
            className="time-label"
            selected={isSelected}
            disabled={isDisabled}
            data-testid={`div-timepicker-cell-label-${cellValue}`}
          >
            {label}
          </Typography>
        </TimeContainer>
      ))}
    </div>
  );
};

const TimePicker = props => {
  const { value, label, placeholder, openByDefault } = props;
  return (
    <CollapsibleFieldContainer
      value={value}
      label={label}
      icon={timeIcon}
      placeholder={placeholder}
      openByDefault={openByDefault}
    >
      <Picker {...props} />
    </CollapsibleFieldContainer>
  );
};

export default TimePicker;

TimePicker.defaultProps = {
  value: null,
  fromTime: '07:00',
  toTime: '20:00',
  interval: '01:00',
  onChange: () => {},
  label: 'Time',
  placeholder: 'Select Time',
  shouldDisableTime: null,
  openByDefault: false,
};

TimePicker.propTypes = {
  value: PropTypes.string,
  fromTime: PropTypes.string,
  toTime: PropTypes.string,
  interval: PropTypes.string,
  onChange: PropTypes.func,
  label: PropTypes.string,
  placeholder: PropTypes.string,
  shouldDisableTime: PropTypes.func,
  openByDefault: PropTypes.bool,
};

Picker.propTypes = TimePicker.propTypes;
