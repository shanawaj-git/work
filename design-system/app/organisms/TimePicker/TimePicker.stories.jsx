/* eslint-disable react/prop-types */

import React, { useState } from 'react';
import StoryWrapper from 'molecules/utility/StoryWrapper';
import TimePicker from './TimePicker';

export default {
  title: 'Components/TimePicker',
  component: TimePicker,
  argTypes: {
    themeType: {
      control: { type: 'select', options: ['DEFAULT', 'MPC', 'AUTOMOTIVE'] },
    },
    children: { table: { disable: true } },
    others: { table: { disable: true } },
  },
};

const defaultProps = {
  value: null,
  fromTime: '07:00',
  toTime: '20:00',
  interval: '01:00',
  onChange: () => {},
  label: 'Time',
  placeholder: 'Select Time',
  shouldDisableTime: null,
  openByDefault: false,
};

const Template = args => {
  const [value, setValue] = useState(args.value);

  return (
    <StoryWrapper themeVariant={args.themeType}>
      <TimePicker {...args} value={value} onChange={setValue} />
    </StoryWrapper>
  );
};

export const BasicTemplate = Template.bind({});
BasicTemplate.args = {
  ...defaultProps,
};

export const TimeDisabledInBetweenTemplate = Template.bind({});
TimeDisabledInBetweenTemplate.args = {
  ...defaultProps,
  shouldDisableTime: time => ['12:00', '13:00'].includes(time),
};

export const OpenByDefault = Template.bind({});
OpenByDefault.args = {
  ...defaultProps,
  openByDefault: true,
};
