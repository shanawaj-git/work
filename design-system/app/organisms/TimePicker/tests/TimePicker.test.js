import React from 'react';
import { render, fireEvent, waitForElement } from 'react-testing-library';
import TimePicker from '../index';

describe('TimePicker', () => {
  const props = {
    value: null,
    fromTime: '07:00',
    toTime: '20:00',
    interval: '01:00',
    onChange: () => {},
    label: 'Time',
    placeholder: 'Select Time',
    shouldDisableTime: null,
    openByDefault: true,
  };
  const testValue = '10:00';

  it('Should render successfully in closed mode', () => {
    const tree = render(<TimePicker {...props} openByDefault={false} />);
    expect(tree).toBeTruthy();
    expect(tree.container).toMatchSnapshot();
  });

  it('Should render successfully in open mode', async () => {
    const tree = render(<TimePicker {...props} />);

    await waitForElement(() =>
      tree.getByTestId(`div-timepicker-cell-container-${testValue}`),
    );
    expect(tree).toBeTruthy();
    expect(tree.container).toMatchSnapshot();
  });

  it("Should invoke 'onChange' on click of a cell with the cell value", async () => {
    const onChange = jest.fn();
    const tree = render(<TimePicker {...props} onChange={onChange} />);
    await waitForElement(() =>
      tree.getByTestId(`div-timepicker-cell-container-${testValue}`),
    );
    const button = tree.getByTestId(
      `div-timepicker-cell-container-${testValue}`,
    );
    fireEvent(
      button,
      new MouseEvent('click', {
        bubbles: true,
        cancelable: true,
      }),
    );
    expect(onChange).toBeCalledWith(testValue);
  });
  it("Should not invoke 'onChange' on click of a disabled cell ", async () => {
    const onChange = jest.fn();
    const disabledValue = '10:00';
    const tree = render(
      <TimePicker
        {...props}
        onChange={onChange}
        shouldDisableTime={time => time === disabledValue}
      />,
    );
    await waitForElement(() =>
      tree.getByTestId(`div-timepicker-cell-container-${testValue}`),
    );
    const button = tree.getByTestId(
      `div-timepicker-cell-container-${disabledValue}`,
    );
    fireEvent(
      button,
      new MouseEvent('click', {
        bubbles: true,
        cancelable: true,
      }),
    );
    expect(onChange).not.toBeCalled();
  });
});
