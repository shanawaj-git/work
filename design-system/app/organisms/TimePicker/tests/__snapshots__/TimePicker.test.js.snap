// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimePicker Should render successfully in closed mode 1`] = `
<div>
  <div
    class="styledComponents__MainContainer-i3ilf2-0 gPVVFI"
  >
    <div
      class="styledComponents__HeaderContainer-i3ilf2-3 ffkpyW"
      data-testid="collapsible-header-container"
    >
      <div
        class="styledComponents__IconContainer-i3ilf2-4 bsvKmv collapsible-header-icon"
      >
        <img
          alt="Select Time"
          class="styledComponents__IconImage-i3ilf2-5 CoppR"
          src="IMAGE_MOCK"
        />
      </div>
      <p
        class="MuiTypography-root MuiTypography-body1 collapsible-header-label css-1reifv3-MuiTypography-root"
      >
        Time
      </p>
      <p
        class="MuiTypography-root MuiTypography-body1 collapsible-header-placeholder text-align-right css-1reifv3-MuiTypography-root"
      >
        Select Time
      </p>
    </div>
  </div>
</div>
`;

exports[`TimePicker Should render successfully in open mode 1`] = `
<div>
  <div
    class="styledComponents__MainContainer-i3ilf2-0 gPVVFI"
  >
    <div
      class="styledComponents__HeaderContainer-i3ilf2-3 ffkpyW"
      data-testid="collapsible-header-container"
    >
      <div
        class="styledComponents__IconContainer-i3ilf2-4 bsvKmv collapsible-header-icon"
      >
        <img
          alt="Select Time"
          class="styledComponents__IconImage-i3ilf2-5 CoppR"
          src="IMAGE_MOCK"
        />
      </div>
      <p
        class="MuiTypography-root MuiTypography-body1 collapsible-header-label css-1reifv3-MuiTypography-root"
      >
        Time
      </p>
      <p
        class="MuiTypography-root MuiTypography-body1 collapsible-header-placeholder text-align-right css-1reifv3-MuiTypography-root"
      >
        Select Time
      </p>
    </div>
    <div
      class="styledComponents__ContentContainer-i3ilf2-2 hOuB"
    >
      <div
        class="styledComponents__Divider-i3ilf2-1 kSSSdW"
      />
      <div
        class="timepicker-container p3 grid grid-cols-3 md:grid-cols-4 lg:grid-cols-6 xl:grid-cols-8 2xl:grid-cols-10"
      >
        <div
          class="styledComponents__TimeContainer-sc-85l0qm-0 biZaIB timepicker-time-cell"
          data-testid="div-timepicker-cell-container-07:00"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 time-label css-1reifv3-MuiTypography-root"
            data-testid="div-timepicker-cell-label-07:00"
          >
            7:00
          </p>
        </div>
        <div
          class="styledComponents__TimeContainer-sc-85l0qm-0 biZaIB timepicker-time-cell"
          data-testid="div-timepicker-cell-container-08:00"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 time-label css-1reifv3-MuiTypography-root"
            data-testid="div-timepicker-cell-label-08:00"
          >
            8:00
          </p>
        </div>
        <div
          class="styledComponents__TimeContainer-sc-85l0qm-0 biZaIB timepicker-time-cell"
          data-testid="div-timepicker-cell-container-09:00"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 time-label css-1reifv3-MuiTypography-root"
            data-testid="div-timepicker-cell-label-09:00"
          >
            9:00
          </p>
        </div>
        <div
          class="styledComponents__TimeContainer-sc-85l0qm-0 biZaIB timepicker-time-cell"
          data-testid="div-timepicker-cell-container-10:00"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 time-label css-1reifv3-MuiTypography-root"
            data-testid="div-timepicker-cell-label-10:00"
          >
            10:00
          </p>
        </div>
        <div
          class="styledComponents__TimeContainer-sc-85l0qm-0 biZaIB timepicker-time-cell"
          data-testid="div-timepicker-cell-container-11:00"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 time-label css-1reifv3-MuiTypography-root"
            data-testid="div-timepicker-cell-label-11:00"
          >
            11:00
          </p>
        </div>
        <div
          class="styledComponents__TimeContainer-sc-85l0qm-0 biZaIB timepicker-time-cell"
          data-testid="div-timepicker-cell-container-12:00"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 time-label css-1reifv3-MuiTypography-root"
            data-testid="div-timepicker-cell-label-12:00"
          >
            12:00
          </p>
        </div>
        <div
          class="styledComponents__TimeContainer-sc-85l0qm-0 biZaIB timepicker-time-cell"
          data-testid="div-timepicker-cell-container-13:00"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 time-label css-1reifv3-MuiTypography-root"
            data-testid="div-timepicker-cell-label-13:00"
          >
            13:00
          </p>
        </div>
        <div
          class="styledComponents__TimeContainer-sc-85l0qm-0 biZaIB timepicker-time-cell"
          data-testid="div-timepicker-cell-container-14:00"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 time-label css-1reifv3-MuiTypography-root"
            data-testid="div-timepicker-cell-label-14:00"
          >
            14:00
          </p>
        </div>
        <div
          class="styledComponents__TimeContainer-sc-85l0qm-0 biZaIB timepicker-time-cell"
          data-testid="div-timepicker-cell-container-15:00"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 time-label css-1reifv3-MuiTypography-root"
            data-testid="div-timepicker-cell-label-15:00"
          >
            15:00
          </p>
        </div>
        <div
          class="styledComponents__TimeContainer-sc-85l0qm-0 biZaIB timepicker-time-cell"
          data-testid="div-timepicker-cell-container-16:00"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 time-label css-1reifv3-MuiTypography-root"
            data-testid="div-timepicker-cell-label-16:00"
          >
            16:00
          </p>
        </div>
        <div
          class="styledComponents__TimeContainer-sc-85l0qm-0 biZaIB timepicker-time-cell"
          data-testid="div-timepicker-cell-container-17:00"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 time-label css-1reifv3-MuiTypography-root"
            data-testid="div-timepicker-cell-label-17:00"
          >
            17:00
          </p>
        </div>
        <div
          class="styledComponents__TimeContainer-sc-85l0qm-0 biZaIB timepicker-time-cell"
          data-testid="div-timepicker-cell-container-18:00"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 time-label css-1reifv3-MuiTypography-root"
            data-testid="div-timepicker-cell-label-18:00"
          >
            18:00
          </p>
        </div>
        <div
          class="styledComponents__TimeContainer-sc-85l0qm-0 biZaIB timepicker-time-cell"
          data-testid="div-timepicker-cell-container-19:00"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 time-label css-1reifv3-MuiTypography-root"
            data-testid="div-timepicker-cell-label-19:00"
          >
            19:00
          </p>
        </div>
        <div
          class="styledComponents__TimeContainer-sc-85l0qm-0 biZaIB timepicker-time-cell"
          data-testid="div-timepicker-cell-container-20:00"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 time-label css-1reifv3-MuiTypography-root"
            data-testid="div-timepicker-cell-label-20:00"
          >
            20:00
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
